nifi-nar-bundles/nifi-gcp-bundle/nifi-gcp-processors/src/main/java/org/apache/nifi/processors/gcp/pubsub/PublishGCPubSub.java

```diff
@@ -213,61 +289,139 @@ public List<ConfigVerificationResult> verify(final ProcessContext context, final
+        }
+        finishBatch(session, flowFileBatch, stopWatch, messageTracker);
+    }
+
+    private ApiFuture<String> publishOneRecord(
+            final ProcessContext context,
+            final FlowFile flowFile,
+            final ByteArrayOutputStream baos,
+            final RecordSetWriter writer,
+            final Record record) throws IOException {
+        baos.reset();
+        writer.write(record);
+        writer.flush();
+        return publishOneMessage(context, flowFile, baos.toByteArray());
+    }
+
+    private ApiFuture<String> publishOneMessage(final ProcessContext context,
+                                                final FlowFile flowFile,
+                                                final byte[] content) {
+        final PubsubMessage message = PubsubMessage.newBuilder()
+                .setData(ByteString.copyFrom(content))
+                .setPublishTime(Timestamp.newBuilder().build())
+                .putAllAttributes(getDynamicAttributesMap(context, flowFile))
+                .build();
+        return publisher.publish(message);
+    }
+
+    private void finishBatch(final ProcessSession session,
+                             final List<FlowFile> flowFileBatch,
+                             final StopWatch stopWatch,
+                             final MessageTracker messageTracker) {
+        try {
+            getLogger().trace("Submit of batch complete, size {}", messageTracker.size());
+            final List<String> messageIdsSuccess = ApiFutures.successfulAsList(messageTracker.getFutures()).get();
+            getLogger().trace("Send of batch complete, success size {}", messageIdsSuccess.size());
+            messageTracker.reconcile(messageIdsSuccess);
+            final String topicName = publisher.getTopicNameString();
+            for (final FlowFileResult flowFileResult : messageTracker.getFlowFileResults()) {
+                final Map<String, String> attributes = new HashMap<>();
+                //attributes.put(MESSAGE_ID_ATTRIBUTE, messageIdFuture.get());  // what to do if using record strategy?
+                attributes.put(TOPIC_NAME_ATTRIBUTE, topicName);
+                final FlowFile flowFile = session.putAllAttributes(flowFileResult.getFlowFile(), attributes);
+                final String transitUri = String.format(TRANSIT_URI_FORMAT_STRING, topicName);
+                session.getProvenanceReporter().send(flowFile, transitUri, stopWatch.getElapsed(TimeUnit.MILLISECONDS));
+                session.transfer(flowFile, flowFileResult.getRelationship());
+                if (flowFileResult.getException() != null) {
+                    getLogger().error("FlowFile send failure", flowFileResult.getException());
                 }
             }
+        } catch (final InterruptedException | ExecutionException e) {
+            session.rollback();

```

Comment:
Session rollback should be removed and instead, FlowFiles should be routed to the appropriate location.

The PubSub API does not appear to support the concept of client-controlled offset handling, which means the message tracker should be used to categorize and route FlowFiles and Records as needed. For the FlowFile strategy, the entire FlowFile could be routed to failure or retry. For the Record strategy, the best option seems to be writing success records and failure records to separate new FlowFiles. That would avoid duplicate messages and also support retrying failed messages on a per-record basis.

Reply:
This makes sense.  I'm going to commit the other feedback, and work this as a separate commit.

Reply:
Done